### PR TITLE
Bump test coverage.

### PR DESCRIPTION
**Pull request by hidmic**
This PR covers:

* init/shutdown API
* context fini API
* node init/fini API
* guard condition init/fini API
* security APIs

to the extent that makes sense using mocks. 

Depends on #760, #761, #762, #763.

### CODE REVIEW

File: rcl/test/rcl/test_context.cpp [hunk 1]
```diff
@@ -170,6 +174,10 @@ TEST_F(CLASSNAME(TestContextFixture, RMW_IMPLEMENTATION), bad_fini) {
   ret = rcl_shutdown(&context);
   EXPECT_EQ(ret, RCL_RET_OK);
 
-  ret = rcl_context_fini(&context);
```
Comment (brawner) [inline rcl/test/rcl/test_context.cpp:173]: Does `inject_on_return` perform the same functionality as the non-mocked function and just return a new different value?

File: rcl/test/rcl/test_security.cpp [hunk 2]
```diff
@@ -253,3 +257,160 @@ TEST_F(TestGetSecureRoot, test_get_security_options) {
...
+  rcl_reset_error();
+}
+
+TEST_F(TestGetSecureRoot, test_rcl_get_secure_root_with_internal_errors) {
+  {
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rcutils_get_env,
+      [&](const char * name, const char ** value) -> const char * {
+        if (strcmp(ROS_SECURITY_ENCLAVE_OVERRIDE, name) == 0) {
+          return "internal error";
```
Comment (brawner) [inline rcl/test/rcl/test_security.cpp:369]: Is this string specifically chosen? Or could it be anything?

File: rcl/test/rcl/test_guard_condition.cpp [hunk 3]
```diff
@@ -202,6 +214,18 @@ TEST_F(
   EXPECT_EQ(RCL_RET_OK, ret);
   ret = rcl_guard_condition_fini(&guard_condition);
   EXPECT_EQ(RCL_RET_OK, ret);
+  // Try normal init and fini, but force an internal error on first try.
+  {
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rcl", rmw_destroy_guard_condition, RMW_RET_ERROR);
+    ret = rcl_guard_condition_init(&guard_condition, &context, default_options);
```
Comment (Blast545) [inline rcl/test/rcl/test_guard_condition.cpp:221]: Shouldnt this init be left outside of the block that mocks the `rmw_destroy_guard_condition`?

File: rcl/test/rcl/test_node.cpp [hunk 4]
```diff
@@ -424,6 +434,18 @@ TEST_F(CLASSNAME(TestNodeFixture, RMW_IMPLEMENTATION), test_rcl_node_life_cycle)
   EXPECT_EQ(RCL_RET_OK, ret);
   ret = rcl_node_fini(&node);
   EXPECT_EQ(RCL_RET_OK, ret);
+  // Try normal init and fini, but force an internal error on first try.
+  {
+    ret = rcl_node_init(&node, name, namespace_, &context, &default_options);
```
Comment (Blast545) [inline rcl/test/rcl/test_node.cpp:439]: Similar comment to the previous

**Comment by Blast545** [comment]
Check PR job before merging

File: rcl/test/rcl/test_security.cpp (hunk 2 above)
Comment (hidmic) [reply rcl/test/rcl/test_security.cpp:369]: Completely arbitrary.

File: rcl/test/rcl/test_context.cpp (hunk 1 above)
Comment (hidmic) [reply rcl/test/rcl/test_context.cpp:173]: Correct.

File: rcl/test/rcl/test_guard_condition.cpp (hunk 3 above)
Comment (hidmic) [reply rcl/test/rcl/test_guard_condition.cpp:221]: It can be outside, it doesn't have to be outside.

File: rcl/test/rcl/test_node.cpp (hunk 4 above)
Comment (hidmic) [reply rcl/test/rcl/test_node.cpp:439]: Same as above.

**Comment by hidmic** [comment]
Alright, using mocks and fault injection I've covered as much as I can reasonably achieve here. I've rebased on top of #768 as that fix is necessary here.

**Comment by hidmic** [comment]
CI up to `rcl`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12124/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7121/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9873/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12029/)

**Comment by brawner** [comment]
Linux retry
[Build Status](https://ci.ros2.org/job/ci_linux/12126/)

**Comment by hidmic** [comment]
@brawner @Blast545 anything else?

File: rcl/test/rcl/test_init.cpp [hunk 5]
```diff
@@ -316,6 +319,76 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_init_and_shutdown
...
+    rcl_reset_error();
+    EXPECT_FALSE(rcl_context_is_valid(&context));
+  }
+
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    ret = rcl_init(test_args.argc, test_args.argv, &init_options, &context);
+
+    int64_t count = rcutils_fault_injection_get_count();
+    rcutils_fault_injection_set_count(RCUTILS_FAULT_INJECTION_NEVER_FAIL);
```
Comment (Blast545) [inline rcl/test/rcl/test_init.cpp:350]: May I ask, why setting this variable here? AFAIK thought the purpose of the `RCUTILS_FAULT_INJECTION_TEST` is to change that variable inside the fault injection test

File: rcl/test/rcl/test_node.cpp [hunk 6]
```diff
@@ -437,6 +425,121 @@ TEST_F(CLASSNAME(TestNodeFixture, RMW_IMPLEMENTATION), test_rcl_node_life_cycle)
...
+  {
+    ret = rcl_node_init(&node, name, namespace_, &context, &options);
+    EXPECT_EQ(RCL_RET_OK, ret);
+    auto mock = mocking_utils::inject_on_return("lib:rcl", rmw_destroy_node, RMW_RET_ERROR);
+    ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+    rcl_reset_error();
+  }
+
+  // Battle test node init/fini.
```
Comment (Blast545) [inline rcl/test/rcl/test_node.cpp:523]: In this test, only init is the one that could fail due to fault injection, right? in that case 
```suggestion
  // Battle test node init.
```

Comment (hidmic) [reply rcl/test/rcl/test_node.cpp]: That is correct! See b400aeea833a7cdb2b2856aea85ed82e3b43beea.

File: rcl/test/rcl/test_init.cpp (hunk 5 above)
Comment (hidmic) [reply rcl/test/rcl/test_init.cpp:350]: This is the current way of temporarily disabling fault injection. I don't want faults to be injected anywhere but within `rcl_init`.

**Comment by hidmic** [comment]
Alright, going in!

File: rcl/test/rcl/test_init.cpp (hunk 5 above)
Comment (Blast545) [reply rcl/test/rcl/test_init.cpp:350]: Yeah, it was clearer to me in the other test

**Comment by hidmic** [comment]
CI up to `rcl`, hopefully the last time:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12132/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7126/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9878/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12035/)

**Comment by hidmic** [comment]
Alright, finally all green! Going in !

**Pull request merged**